cedar-policy-validator/src/lib.rs

```diff
@@ -206,6 +267,205 @@ impl Validator {
+    ) -> (EntityDerefLevel, Option<EntityDerefLevelViolation>) {
+        let p = v.into_iter().min_by(|(l1, _), (l2, _)| l1.cmp(l2));
+        match p {
+            Some(p) => p.clone(),
+            None => (EntityDerefLevel { level: 0 }, None),
+        }
+    }
+
+    #[cfg(feature = "level-validate")]
+    /// Walk the type-annotated AST and compute the used level and possible violation
+    /// Returns a tuple of `(actual level used, optional violation information)`
+    fn check_entity_deref_level_helper<'a>(
+        &'a self,
+        e: &cedar_policy_core::ast::Expr<Option<crate::types::Type>>,
+        max_allowed_level: &EntityDerefLevel,
+        policy_id: &PolicyID,
+    ) -> (EntityDerefLevel, Option<EntityDerefLevelViolation>) {
+        use crate::types::{EntityRecordKind, Type};
+        use cedar_policy_core::ast::ExprKind;
+        match e.expr_kind() {
+            ExprKind::Lit(_) => (
+                EntityDerefLevel { level: 0 }, //Literals can't be dereferenced
+                None,
+            ),
+            ExprKind::Var(_) => (max_allowed_level.clone(), None), //Roots start at `max_allowed_level`
+            ExprKind::Slot(_) => (EntityDerefLevel { level: 0 }, None), //Slot will be replaced by Entity literal so treat the same
+            ExprKind::Unknown(_) => (
+                EntityDerefLevel { level: 0 }, //Can't dereference an unknown
+                None,
+            ),
+            ExprKind::If {
+                test_expr,
+                then_expr,
+                else_expr,
+            } => {
+                let es = [test_expr, then_expr, else_expr];
+                let v: Vec<(EntityDerefLevel, Option<_>)> = es
+                    .iter()
+                    .map(|l| self.check_entity_deref_level_helper(l, max_allowed_level, policy_id))
+                    .collect();
+                Self::min(v)
+            }
+            ExprKind::And { left, right } | ExprKind::Or { left, right } => {
+                let es = [left, right];
+                let v: Vec<(EntityDerefLevel, Option<_>)> = es
+                    .iter()
+                    .map(|l| self.check_entity_deref_level_helper(l, max_allowed_level, policy_id))
+                    .collect();
+                Self::min(v)
+            }
+            ExprKind::UnaryApp { arg, .. } => {
+                self.check_entity_deref_level_helper(arg, max_allowed_level, policy_id)
+            }
+            // `In` operator decrements the LHS only
+            ExprKind::BinaryApp { op, arg1, arg2 } if op == &BinaryOp::In => {
+                let mut lhs =
+                    self.check_entity_deref_level_helper(arg1, max_allowed_level, policy_id);
+                let rhs = self.check_entity_deref_level_helper(arg2, max_allowed_level, policy_id);
+                lhs = (lhs.0.decrement(), lhs.1);
+                let new_level = Self::min(vec![lhs, rhs]).0;
```

Comment:
I'm not sure why we're looking for min here? 
> The rules for expression e1 in e2 (line (3) in Fig. 7)(a) in the paper) are extended to require that the entity type of e1 is E1(n) where n > 0. This requirement indicates that e1 must be an entity from which you can dereference at least one level (to access its ancestors).

Could you explain?

Reply:
I _think_ this is correct, it's just phrased very different from the type system version. arg2 can be zero, because it checks for being less than zero, which only happens due to decrements.

Reply:
ISTM the `min` is intended to handle situations where there may be a levels violation on the RHS of the `in`. Something like `principal in resource.foo.bar` validated at level 0 might get reported as a level violation at level 2 due to `resource.foo.bar` rather than at level 1 due to `principal in ...`. 

I asked about the same code here: https://github.com/cedar-policy/cedar/pull/1146#discussion_r1783398534

Probably worth another test demonstrating the case where the `min` value is determined b `arg2`, but `arg1` level is also less than 0.